apps/docs/refactor.md

```diff
@@ -0,0 +1,565 @@
+
+Here is how the above files might look if we wanted to expose only step
+incrementation, but not step decrementation.
+
+**`src/stepper/redux/actionTypes.js`**
+
+Action types are prefixed with the name of the feature to avoid collisions with
+other libraries and their reducers.
+
+```js
+export const INCREMENT_STEP = "stepper/INCREMENT_STEP";
+export const DECREMENT_STEP = "stepper/DECREMENT_STEP";
+```
+
+**`src/stepper/redux/reducer.js`**
+
+The reducer for the feature library should operate on a subtree of the global
+state tree, rather than on the entire state tree. Note that this library acts
+like a singleton in the sense that there is only one corresponding state object
+for the library on the page at any given time.
+
+```js
+import * as actionTypes from './actionTypes';
+
+export default reducer(state={step: 1}, action) {
+  switch (action.type) {
+    case actionTypes.INCREMENT_STEP:
+      return {step: state.step + 1};
+    case actionTypes.DECREMENT_STEP:
+      return {step: state.step - 1};
+    default:
+      return state;
+  }
+}
+```
+
+**`src/stepper/redux/actionCreators.js`**
+
+Action creators should make sure to generate actions with types that are
+globally unique on the page.
+
+```js
+import * as actionTypes from './actionTypes';
+
+export function incrementStep() {
+  return {type: actionTypes.INCREMENT_STEP}
+}
+
+export function decrementStep() {
+  return {type: actionTypes.DECREMENT_STEP}
+}
+```
+
+**`src/stepper/redux/selectors.js`**
+
+Selector functions for the feature must be "configured" before they can be used
+so that they know how to access the part of the global state tree with the data
+they are interested in. This is done using the `setRootSelector` function.
+
+```js
+let getRoot = () => throw new Error("rootSelector not set. Did you forget to call configure?");
+
+export function setRootSelector(rootSelector) {
+  getRoot = rootSelector;
+}
+
+export function getStep(state) {
+  return getRoot(state).step;
+}
+```
+
+**`src/stepper/index.js`**
+
+The feature library's `index.js` file exposes just the action types, action
+creators, and selectors that it expects other libraries to use. This makes it
+easier to explicitly limit how a library can be used, reducing the chance of
+unintentional coupling and spaghetti code.
+
+`index.js` also exposes a `configure` function that is used to inject the root
+selector into the feature library from above, which removes coupling between the
+feature library and the exact format of the global redux store.
+
+```js
+import reducer from './redux/reducer';
+import {INCREMENT_STEP} from './redux/actionTypes';
+import {incrementStep} from './redux/actionCreators';
+import {getStep, setRootSelector} from './redux/selectors';
+
+export reducer;
+export const actionTypes = {INCREMENT_STEP}
+export const actions = {incrementStep};
+export const selectors = {getStep};
+
+export function configure(config) {
+  setRootSelector(config.getState);
```

Comment:
But `setRootSelector` mutates a 'static' variable so both instances would end up with the same root selector.  You'd have to do something much more elaborate to instantiate libraries with different root selectors.